### PR TITLE
CiviEventInspector - Add util for displaying hook metadata

### PR DESCRIPTION
**Pull request by totten**
This is meant to facilitate better code-generators and documentation-generators. Key features:

 * Adds a new class, `Civi\Core\CiviEventInspector`.
 * Adds a new hook, `hook_civicrm_eventDefs`. Extensions may use this to document their hooks.
 * Implements the hook in core (to report on `civicrm-core` hooks and events)
 * If you follow the practice of declaring static hook stubs (like `civicrm-core` does), then use `addStaticStubs()` to assimilate of them.

(Description updated to reflect current class/hook names.)

### CODE REVIEW

**Comment by totten** [comment]
On second thought, if the event dispatcher is [the canonical representation of hooks](https://github.com/civicrm/civicrm-core/pull/9949), then perhaps I should rename it to `EventInspector` and include metadata about all events (regardless of whether they're named `hook_foo`). That's a small number of additional events, but it'd more complete.

**Comment by totten** [comment]
@xurizaemon @JohnFF I've updated this to report all events, including both hooks and oddballs.

To get a list of hooks, you could run:

```
cv ev '$i = new \Civi\Core\CiviEventInspector(); return $i->find("/^hook_/");'
```

File: CRM/Admin/Page/APIExplorer.php [hunk 1]
```diff
@@ -184,9 +184,14 @@ private static function getDocBlock($entity, $action) {
...
-  private static function formatDocBlock($text) {
+  public static function formatDocBlock($text) {
+    // Normalize #leading spaces.
+    $lines = explode("\n", $text);
+    $lines = preg_replace('/^ +\*/', ' *', $lines);
+    $text = implode("\n", $lines);
+
     // Get rid of comment stars
-    $text = str_replace(array("\n * ", "\n *\n", "\n */\n", "/**\n"), array("\n", "\n\n", '', ''), $text);
+    $text = str_replace(array("\n * ", "\n * ", "\n *\n", "\n */\n", "/**\n"), array("\n", "\n\n", '', ''), $text);
```
Comment (totten) [inline CRM/Admin/Page/APIExplorer.php]: Oops, that last bit's a mistake. I'll squash it out.

**Comment by eileenmcnaughton** [comment]
@xurizaemon this is out of WIP now & you seem to have some familiarity with it - it looks OK to me - can you confirm you think it is good to merge?

**Comment by xurizaemon** [comment]
It seems like it will be relatively safe since nothing of significance is using it. ShowAllTheHooks is using it (see prev commit from fuzionnz/contrib.showallthehooks) but that's hardly an extensive test ... it does work, and it gives an interface that informs for available hooks and events, which is great.

**Comment by eileenmcnaughton** [comment]
I think given the scope of this I agree with @xurizaemon and am good to merge

**Pull request merged**